File: 2448-minimum-cost-to-make-array-equal/2448-minimum-cost-to-make-array-equal.cpp
class Solution {
public:
    long long go(vector<int>&nums,vector<int>&cost,long long x){
        long long ans =0;
        for(int i=0;i<nums.size();++i){
            ans+=(cost[i]*(abs(nums[i]-x)));
        }
        return ans;
    }
    long long minCost(vector<int>& nums, vector<int>& cost) {
        long long low = *min_element(begin(nums),end(nums));
        long long high = *max_element(begin(nums),end(nums));
        long long res = go(nums,cost,low);
        while(low<high){
            long long mid = (high+low)/2;
            long long c1 = go(nums,cost,mid),c2=go(nums,cost,mid+1);
            res = min(c1,c2);
            if(c1<c2)
                high = mid;
            else
                low = mid+1;
        }
        return res;
    }
};

File: 2448-minimum-cost-to-make-array-equal/2448-minimum-cost-to-make-array-equal.py
class Solution:
    def minCost(self, nums: List[int], cost: List[int]) -> int:
        def go(x):
            ans = 0
            nonlocal cost
            nonlocal nums
            for i in range(len(nums)):
                ans +=(cost[i]*(abs(nums[i]-x)))
            return ans
        low = min(nums)
        high = max(nums)
        res = go(low)
        while(low<high):
            mid = (low+high)//2
            c1 = go(mid)
            c2 = go(mid+1)
            res = min(c1,c2)
            if c1<c2 :
                high = mid
            else :
                low = mid+1
        return res